### PR TITLE
Garbage collection, asset delivery, icon2html revolution, and general fixes (515)

### PR DESCRIPTION
**Pull request by cadyn**
OLD:
Piggybacks off of https://github.com/VOREStation/VOREStation/pull/15320 due to the need for refcount function. Tested deletion of hundreds of mobs and seems to run faster than previous.

UPDATE:
Okay, so we've been running 515 on torch for a while and it's seemed mostly stable other than some issues with TGUI. I've been slowly working on getting GC better and better, and finally got to the point where I was comfortable removing the harddelete hint from mobs and well the results speak for themselves:
image
Previously this many deletions would have taken ages, anywhere from 10 minutes to an hour depending on server load. It's done in seconds. 

Asset delivery:
This is bundled in now, ported directly from TGStation this should improve the current system for delivering assets(sprite sheets, js, css, html, images, etc.) to be hopefully more reliable. Also includes more settings so that things like preloading can be enabled as well as the option to serve assets from a webserver to improve transfer speeds (untested currently).

Observer datums:
Fucking axed the shit out of these. They were a much worse version of COMSIGs that used global lists. Replaced all of them with COMSIGs that should preserve their previous functionality. This change was related to GC optimization.

Icon2html revolution:
I have to thank Kash, Drulikar, and Selis from the Chomp discord for helping me figure out what needed to be done here. For the longest time there were issues with TGUI on 515 and switching all of the chat icons to use Icon2html seems to have resolved that. There are a few bicons laying around still because they were hidden behind procs that couldn't get a client to send assets to, but they shouldn't cause too much of an issue.

### CODE REVIEW

**Comment by Kashargul** [comment]
might not want to have the custom item config file in there.

**Comment by cadyn** [comment]
dreamchecker is angy because it doesn't recognize refcount from 515

**Comment by cadyn** [comment]
Yea I'm not even going to touch the AST code. I don't know why it's angry because of that, but it's not my problem.

**Comment by cadyn** [comment]
Just to be clear, I'm not bothering with dreamchecker. It's giving annotations for files that I didn't touch, and that seem to compile fine. I've already spent a hell of a lot of time getting all of this working in this first place.

**Comment by ItsSelis** [comment]
Would merging #15727 already solve the /icon stuff in this PR?
That would be at least somewhat of a atomization that makes reviewing a lot of files a bit easier (as most likely a lot of them are just the change from /icon to icon2html)

**Comment by Kashargul** [comment]
Nope, we can close the /icon PR, this one builds on it. Going a step further by replacing the really outdated base64 lookup list which is simply horrible in performance.

**Comment by ItsSelis** [comment]
And yeah as Kash said the linter errors appear because of the newer byond/dreamchecker version as "operator" is not really valid, or rather is a reserved keyword in byond.

(Also the reason why StrongDMM doesn't work here but does on downstream).

**Comment by ItsSelis** [comment]
Still any kind of atomization, if possible, is pretty good to have faster reviews of smaller changes that can stand on their own.

**Comment by cadyn** [comment]
> Still any kind of atomization, if possible, is pretty good to have faster reviews of smaller changes that can stand on their own.

I do understand the benefits of atomizing something like this, and I may have even done so if I could go back. As it stands, it was easiest for me to get all of this in one PR, though. All of these changes do connect to each other in odd ways:

- 515 is required for GC update
- nuking observers & some other various patches make GC run faster
- Asset cache update & icon2html revolution are necessary to get 515 working properly which ties back to GC

Basically, yeah, I see why you would want some of the more time sensitive of these changes to be atomized so they can make their way into production sooner, especially for downstreams. As it stands, if you want them though, I would recommend downporting this PR (as I've already done on torch, which-- it might be easier to upport the downport than to downport this) and test-merging if the performance fixes are considered time-sensitive enough that you do not want to wait until 515 gets going on virgo.

**Comment by cadyn** [comment]
Synchronized with https://github.com/CHOMPStation2/CHOMPStation2/pull/7783 to match the latest changes we've made as we've been testing this on both CHOMPstation and Torchstation

**Comment by ItsSelis** [comment]
These 3 commits (iirc) were still pretty new and have to be brought over still (just as a note as the rest is getting merged)

https://github.com/CHOMPStation2/CHOMPStation2/pull/7783/commits/aa199be4db01630e40962a5e41fcc11020af3e1c
https://github.com/CHOMPStation2/CHOMPStation2/pull/7783/commits/6fedb6074682c3a99e6bdc5632c59a1851c27a5a
https://github.com/CHOMPStation2/CHOMPStation2/pull/7783/commits/d7c8e3f5eac2820968b1fea4c1b0d1b2d4a0d515

**Comment by cadyn** [comment]
Fully synced up to the merged PR

**Comment by Heroman3003** [comment]
Conflicts popped up.

**Comment by cadyn** [comment]
> Conflicts popped up.

resolved

**Comment by cadyn** [comment]
Integration test failure is due to active edges, unrelated to PR.

**Comment by cadyn** [comment]
Upport of https://github.com/CHOMPStation2/CHOMPStation2/pull/7894 which adds some important optimizations ontop of the existing content of this PR.
